### ISSUE TITLE
"safeStorage API is not available" persists

### ISSUE THREAD

**Issue by kibibytium** [tutao/tutanota#7257]
The issue #7077 seems to persist still on versions at least 234.240627.0 and later.

## Test notes
- [ ] Ensure there is no config folder for the app (e.g. no `~/.config/tutanota-desktop-test` dir if a test build).
- [ ] Run the app with the `--password-store=basic` command-line argument.
- [ ] Verify you get a warning telling you that basic_text is insecure, but the app still opens without a safeStorage error.
- [ ] Verify you can log in.
- [ ] Restart the app with the same arguments.
- [ ] Verify you can still log in.

**Comment by remi-sc** [comment]
Just bumping this in case someone thinks its been fixed: It wasn't. still a problem in version 246.
There's also another closed issue about it here: https://github.com/tutao/tutanota/issues/6265

I'm stuck on version 220 for a long time now because of this.
(And I never even save my credentials - so should be no need to rely on any storage for it...)

**Comment by remi-sc** [comment]
@charlag 

This is not fixed, https://github.com/tutao/tutanota/releases/tag/tutanota-desktop-release-252.241122.0 suggests it was.

Here's a log from the `252.241122.0` appimage after making sure tuta process is dead and removing `chromium` and `tutanota-desktop` from `.config`

```
% kill -9  $(pidof tutanota-desktop)
% rm -rf .config/chromium
% rm -rf .config/tutanota-desktop  
% ./tutanota-desktop-linux.AppImage 
the monkey has been patched pid: 866120
(node:866120) [DEP0005] DeprecationWarning: Buffer() is deprecated due to security and usability issues. Please use the Buffer.alloc(), Buffer.allocUnsafe(), or Buffer.from() methods instead.
(Use `tutanota-desktop --trace-deprecation ...` to show where the warning was created)
Creating tables...
Current credentials version: 1
Tables created successfully!
version:   252.241122.0
resolving key... tutanota-vault
config up to date
Failed to resolve/generate key:  tutanota-vault K [DeviceStorageUnavailableError]: safeStorage API is not available
    at Pv.assertAvailable (/tmp/.mount_tutanoxoRUAQ/resources/app.asar/desktop/DesktopMain.js:11:190075)
    at async Pv.getPassword (/tmp/.mount_tutanoxoRUAQ/resources/app.asar/desktop/DesktopMain.js:11:188958)
    at async zO.fetchKey (/tmp/.mount_tutanoxoRUAQ/resources/app.asar/desktop/DesktopMain.js:11:207468)
    at async zO.fetchOrGenerateKey (/tmp/.mount_tutanoxoRUAQ/resources/app.asar/desktop/DesktopMain.js:11:207195)
[DesktopMain] getting dictionaries from: https://app.tuta.com/desktop/dictionaries/
[2024-11-25T20:45:39.487Z] Create tray:true
[SSEFacade] last missed notification check: { lastMissedNotificationCheckTime: null }
[ApplicationWindow] webAssetsPath:  /tmp/.mount_tutanoxoRUAQ/resources/app.asar
default mailto handler: false
Webapp ready
[SSEFacade] No SSE info
[2024-11-25T20:45:39.561Z] [ElectronUpdater] INFO: Checking for update
[2024-11-25T20:45:39.561Z] [ElectronUpdater] INFO: checking-for-update
[DesktopMain] spellcheck-dictionary-initialized en-US
/home/user/.local/share/applications/tutanota-desktop.desktop does not exist, checking for permission to ask for permission...
[DesktopMain] getting dictionaries from: https://app.tuta.com/desktop/dictionaries/
[DesktopMain] spellcheck-dictionary-initialized en-US
[2024-11-25T20:45:39.900Z] [ElectronUpdater] INFO: Update for version 252.241122.0 is not available (latest version: 252.241122.0, downgrade is disallowed).
[2024-11-25T20:45:39.900Z] [ElectronUpdater] INFO: update not available: {
  version: '252.241122.0',
  files: [
    {
      url: 'tutanota-desktop-linux.AppImage',
      sha512: 'JAwo0sjLNKP4kBZGhbJN8gZ4msDTropYUvbCeMWnU4Dg3kWaILmVJBvOvt/s1drom3R3yFYSfJY1RPJb1NYj+Q==',
      size: 128265828,
      blockMapSize: 134776
    }
  ],
  path: 'tutanota-desktop-linux.AppImage',
  sha512: 'JAwo0sjLNKP4kBZGhbJN8gZ4msDTropYUvbCeMWnU4Dg3kWaILmVJBvOvt/s1drom3R3yFYSfJY1RPJb1NYj+Q==',
  releaseDate: '2024-11-22T17:11:07.586Z',
  signature: 'MQmfLth32Nb8dU8QnTmgW+JJyTeW5XcfxVenCpDcfGznqTlCt2OwrCSVGacolrGYBtjETP3U6N8EeJOXa8m5J8Uarz6N9+hDGidUp9VG9vNec/xMMcD7i+iNPKhmLjw+2jD3TKTMcoLJGfucGEp9juoIVJkB4UhLHsIiMXck++lNE38WQiuJF4zVQ3geQmCoA051RGk7/GgTH9VaZdlwW0eUfLW3Nr6099o4nwjo85DEf1yid8WJx9lDfOKwFdItmMVfq45TaXcWl/Yw/7K6ZJQPIqcmQtAmN03+Pklj2m72OFbHvHyzDLCyuFkRv+OuQBuqflkQ+v4N8ohYJ6a32Q=='
}
[866120:1125/214543.166740:ERROR:browser_main_loop.cc(279)] GLib-GObject: ../glib/gobject/gsignal.c:2684: instance '0x186c024f3e40' has no handler with id '2499'
[OfflineDbRefCounter] Deleting db for user <<SNIP>>
deleting db for <<SNIP>>
[OfflineDbRefCounter] Deleted db for user <<SNIP>>
[PerWindowSqlCipherFacade] open db for <<SNIP>>
[2024-11-25T20:45:56.355Z] [WorkerSqlCipher] started sqlcipher-worker-1
[2024-11-25T20:45:56.406Z] [sqlcipher-worker-1] set up sql cipher done
resolving key... tutanota-credentials
Failed to resolve/generate key:  tutanota-credentials K [DeviceStorageUnavailableError]: safeStorage API is not available
    at Pv.assertAvailable (/tmp/.mount_tutanoxoRUAQ/resources/app.asar/desktop/DesktopMain.js:11:190075)
    at async Pv.getPassword (/tmp/.mount_tutanoxoRUAQ/resources/app.asar/desktop/DesktopMain.js:11:188958)
    at async zO.fetchKey (/tmp/.mount_tutanoxoRUAQ/resources/app.asar/desktop/DesktopMain.js:11:207468)
    at async zO.fetchOrGenerateKey (/tmp/.mount_tutanoxoRUAQ/resources/app.asar/desktop/DesktopMain.js:11:207195)
    at async kF.encryptUsingKeychain (/tmp/.mount_tutanoxoRUAQ/resources/app.asar/desktop/DesktopMain.js:25:16794)
    at async aD.getOrCreateCredentialEncryptionKey (/tmp/.mount_tutanoxoRUAQ/resources/app.asar/desktop/DesktopMain.js:11:212682)
    at async aD.setCredentialEncryptionMode (/tmp/.mount_tutanoxoRUAQ/resources/app.asar/desktop/DesktopMain.js:11:211726)
    at async ipc (/tmp/.mount_tutanoxoRUAQ/resources/app.asar/desktop/DesktopMain.js:11:221985)

```

**Comment by remi-sc** [comment]
hmm, or rather it seems this is kind of fixed: The client does not fallback to basic storage by itself after failing to find the API,
explicit `--password-store=basic` is required to login. 

I assume this is not intended. There is no useful error message displayed that would suggest enabling this flag, An average user will most likely simply assume the desktop client is broken. (which is nothing new to be honest)

**Comment by remi-sc** [comment]
I am not using KDE or gnome, I am using xfce without any of the libsecret safestorage backend packages. 

On Arch `libsecret` has a few potential ones: [gnome-keyring](https://archlinux.org/packages/extra/x86_64/gnome-keyring/), [keepassxc](https://archlinux.org/packages/extra/x86_64/keepassxc/), [kwallet5](https://archlinux.org/packages/extra/x86_64/kwallet5/), [kwallet](https://archlinux.org/packages/extra/x86_64/kwallet/)

I don't use any of them.
To reproduce: its probably as simple as `pacman -R gnome-keyring keepassxc kwallet5 kwallet`
(although I imagine that this may mean removing whole KDE if it's not a soft dependency)

**Comment by remi-sc** [comment]
Thank you so much for explaining, From my limited user perspective I'd say anything that would avoid a situation when I launch and am presented with an opaque error preventing me from logging in at all would be an improvement.

> The problem is that, if you are running the app for the first time, we have no idea if you want to set an app password

Seems that it's not limited to first login either. even if I:
- open on a clean `.config` from scratch with `--password-store=basic` and login
- setup a app password
- exit gracefully
- relaunch without `--password-store=basic`
- input the app password into a prompt
I still get: "could not access secret storage" and cant login

**Comment by charlag** [comment]
I think we could do two things:
 - remember which backend was used for storing secrets and error out if there's downgrade happening
 - let users know that plaintext backend was used and that it's not super safe

I am a bit on a fence about the second one, I feel like if you are in a situation where you don't have a DE, you are already an advanced user. Maybe just a console message for the second case is enough.

**Comment by remi-sc** [comment]
Another option that I don't know why isn't available: don't store credentials anywhere outside the client memory - requiring providing primary credentials on every login. Up to version 220 that's what I was doing and I didn't have to even think about some backend library storing things less securely. (unless I was misunderstanding it for the longest time and tuta was happily storing things unencrypted without me knowing)

**Comment by charlag** [comment]
Hi, we did change the credentials system for multiple reasons:
 - the previous situation was confusing and went against user expectations. Some data was stored locally (e.g. alarms, push notifications identifier) but was not tied to credentials. Usually you log in into the desktop/mobile apps only once, previous login system was designed in different constraints, primarily for the web app and without local authentication.
 - going through the full login process every time discourages using secure password and 2FA
 - it introduced a lot of development overhead to always consider the case where local data is not stored
 - local per-app data security on desktop is not part of the threat model for most users

We believe that app password effectively addresses all of these issues. We still try to make sure that credentials storage is present as it is additional layer of security and it also "just works" on most desktop setups.

Even prior to the change of how credentials are stored the safe storage was used and some functionality would not work without it so it is mostly orthogonal to the credentials issue.
